### PR TITLE
Massively improve the performance of <system_error> and remove constexpr workarounds.

### PR DESCRIPTION
**Pull request by BillyONeal**
This is a reopening of https://github.com/microsoft/STL/pull/398 with things fixed.

Removes  VSO-406237 "C1XX does not do constant initialization where the standard requires" workarounds.

* Introduce `_Immortalize_memcpy_image` which avoids the need for synchronization when constructing the error categories. See the big comment block for explanation of how we got here. For compilers which let us declare instances with the empty virtual destructor as constexpr variables (currently only clang in C++20 mode), we can avoid the workaround.
* Turn on `/Zc:threadSafeInit-` and warning 4640 in most tests to ensure we aren't emitting thread-local initializers in `<system_error>`.
* Move `_Immortalizer` and friends to the one remaining caller in the `exception_ptr` implementation.
* Remove `_Generic_error_category` from the base class hierarchies of error categories to workaround DevCom-928781.

# Checklist

Be sure you've read README.md and understand the scope of this repo.

If you're unsure about a box, leave it unchecked. A maintainer will help you.

- [x] Identifiers in product code changes are properly `_Ugly` as per
  https://eel.is/c++draft/lex.name#3.1 or there are no product code changes.
- [x] The STL builds successfully and all tests have passed (must be manually
  verified by an STL maintainer before automated testing is enabled on GitHub,
  leave this unchecked for initial submission).
- [x] These changes introduce no known ABI breaks (adding members, renaming
  members, adding virtual functions, changing whether a type is an aggregate
  or trivially copyable, etc.).
- [x] These changes were written from scratch using only this repository,
  the C++ Working Draft (including any cited standards), other WG21 papers
  (excluding reference implementations outside of proposed standard wording),
  and LWG issues as reference material. If they were derived from a project
  that's already listed in NOTICE.txt, that's fine, but please mention it.
  If they were derived from any other project (including Boost and libc++,
  which are not yet listed in NOTICE.txt), you *must* mention it here,
  so we can determine whether the license is compatible and what else needs
  to be done.

### CODE REVIEW

File: stl/inc/system_error [hunk 1]
```diff
@@ -6,14 +6,16 @@
 #pragma once
 #ifndef _SYSTEM_ERROR_
 #define _SYSTEM_ERROR_
-#include <yvals_core.h>
+#include <yvals.h>
```
Comment (BillyONeal) [inline stl/inc/system_error]: @CaseyCarter do you know why this very-non-Core header said `<yvals_core.h>`?

File: stl/src/excptptr.cpp [hunk 2]
```diff
@@ -37,6 +37,51 @@ extern "C" _CRTIMP2 void* __cdecl __AdjustPointer(void*, const PMD&); // defined
...
+    template <class _Ty>
+    int __stdcall _Immortalize_impl(void*, void* _Storage_ptr, void**) noexcept {
+        // adapt True Placement New to _Execute_once
+        ::new (_Storage_ptr) _Ty();
+        return 1;
+    }
+
+    template <class _Ty>
+    _Ty& _Immortalize() { // return a reference to an object that will live forever
+        static_assert(sizeof(void*) == sizeof(once_flag), "TRANSITION, VSO-406237");
```
Comment (BillyONeal) [inline stl/src/excptptr.cpp]: This can be upgraded to real once_flag.

**Comment by barcharcraz** [comment]
was the LLVM submodule bump intended?

**Comment by BillyONeal** [comment]
> was the LLVM submodule bump intended?

Yes, to get https://github.com/llvm/llvm-project/commit/e48849a2404128175df25168f961a83d6c0a901e

File: stl/inc/system_error [hunk 3]
```diff
@@ -6,14 +6,17 @@
...
-#include <yvals_core.h>
+#include <yvals.h>
 #if _STL_COMPILER_PREPROCESSOR
 #include <__msvc_system_error_abi.hpp>
 #include <cerrno>
 #include <cstdlib>
 #include <stdexcept>
-#include <xcall_once.h>
 #include <xerrc.h>
+#include <xcall_once.h>
```
Comment (StephanTLavavej) [inline stl/inc/system_error]: This isn't sorted order; clang-format should fail.

File: stl/inc/future [hunk 4]
```diff
@@ -153,11 +153,9 @@ private:
 };
 
 // CLASS _Future_error_category
```
Comment (StephanTLavavej) [inline stl/inc/future]: Outdated comment.

File: stl/inc/system_error [hunk 5]
```diff
@@ -466,11 +472,9 @@ public:
 };
 
 // CLASS _Iostream_error_category
```
Comment (StephanTLavavej) [inline stl/inc/system_error]: Outdated comment.

File: stl/inc/system_error [hunk 6]
```diff
@@ -520,16 +522,84 @@ public:
...
+// Moreover, we need to be very conservative in the runtime support we request. Thus, we have these constraints:
+//
+// * can't use magic statics in standard modes, because that would inject a .TLS section into all binaries using 
+//   <system_error> and would likely put borderline programs over the TLS slot count limit
+// * can't declare the error_category as an ordinary constexpr variable for most compilers, because error_category
+//   has a virtual destructor (TRANSITION, __cpp_constexpr_dynamic_alloc)
+// * can't declare the error_category as an ordinary non-constexpr variable even with a constexpr constructor, because
+//   the compiler will emit code to destroy it which invalidates its use in these DLL shutdown scenarios
+//
+// As a result, we are left a somewhat unfortunate workaround: We create an atomic<uintptr_t> array to store the
```
Comment (StephanTLavavej) [inline stl/inc/system_error]: Would sound better as "left with", or "we use", or "we have".

File: stl/inc/system_error [hunk 7]
```diff
@@ -520,16 +522,84 @@ public:
...
+//   error_category instance, test if the first atomic is nonzero (acquire), and if so, we know we have formed the
+//   instance and can return a reinterpreted pointer to that storage. If the first atomic is zero, we write all except
+//   the first atomic (relaxed), then write the first one as a store-release. (The non-first values are transferred
+//   to other threads in the release sequence).
+//
+// Acknowledged undefined and implementation defined behavior happening here:
```
Comment (StephanTLavavej) [inline stl/inc/system_error]: Hyphenate implementation-defined as an adjective.

File: stl/inc/system_error [hunk 8]
```diff
@@ -520,16 +522,84 @@ public:
...
+// * There is a data race when filling in the values other than the first atomic; this is OK on all hardware we target
+//   because the racing threads are all writing identical values.
+// * We are reaching into the memory layout atomic<uintptr_t>[N] and assuming we can interpret that as some other type.
```
Comment (StephanTLavavej) [inline stl/inc/system_error]: I would expect "layout of".

Comment (StephanTLavavej) [inline stl/inc/system_error]: "interpret" should probably be "reinterpret".

File: stl/inc/system_error [hunk 9]
```diff
@@ -520,16 +522,84 @@ public:
...
+    return _Static._Storage;
+}
+#pragma warning(pop)
+#elif defined(__cpp_constexpr_dynamic_alloc)
+template <class _Ty>
+_NODISCARD const _Ty& _Immortalize_memcpy_image() noexcept {
+    constexpr _Ty _Static;
+    return _Static._Storage;
+}
+#else
```
Comment (StephanTLavavej) [inline stl/inc/system_error]: A comment here would be nice.

File: stl/inc/system_error [hunk 10]
```diff
@@ -520,16 +522,84 @@ public:
...
+template <class _Ty>
+_NODISCARD const _Ty& _Immortalize_memcpy_image() noexcept {
+    // return reference to a memcpy'd default-initialized _Ty
+    // pre: A default-initialized _Ty sets the first pointer-sized field to nonzero
+    constexpr size_t _Pointer_count = sizeof(_Ty) / sizeof(uintptr_t);
+    static atomic<uintptr_t> _Storage[_Pointer_count];
+    static_assert(sizeof(_Storage) == sizeof(_Ty), "Bad storage size");
+    static_assert(alignof(decltype(_Storage)) >= alignof(_Ty), "Bad alignment assumptions");
+    if (_Storage[0].load(_STD memory_order_acquire) != 0) {
```
Comment (StephanTLavavej) [inline stl/inc/system_error]: We're in `std` and `memory_order_acquire` is not a function, so this should be unqualified.

File: stl/inc/system_error [hunk 11]
```diff
@@ -520,16 +522,84 @@ public:
...
+        return reinterpret_cast<_Ty&>(_Storage);
+    }
+
+    const _Ty _Target;
+    const auto _Target_iter = reinterpret_cast<const uintptr_t*>(_STD addressof(_Target));
+    _CSTD memcpy(_Storage + 1, _Target_iter + 1, sizeof(_Ty) - sizeof(uintptr_t));
+    _Storage[0].store(_Target_iter[0], _STD memory_order_release);
```
Comment (StephanTLavavej) [inline stl/inc/system_error]: Ditto unqualified `memory_order_release`.

File: stl/src/excptptr.cpp [hunk 12]
```diff
@@ -37,6 +37,37 @@ extern "C" _CRTIMP2 void* __cdecl __AdjustPointer(void*, const PMD&); // defined
 using namespace std;
 
 namespace {
+    #ifdef _M_CEE_PURE
```
Comment (StephanTLavavej) [inline stl/src/excptptr.cpp]: There are multiple clang-format issues causing the checks to fail.

File: stl/src/excptptr.cpp [hunk 13]
```diff
@@ -37,6 +37,37 @@ extern "C" _CRTIMP2 void* __cdecl __AdjustPointer(void*, const PMD&); // defined
...
+        // adapt True Placement New to _Execute_once
+        ::new (_Storage_ptr) _Ty();
+        return 1;
+    }
+
+    template <class _Ty>
+    _Ty& _Immortalize() { // return a reference to an object that will live forever
+        static once_flag _Flag;
+        alignas(_Ty) static unsigned char _Storage[sizeof(_Ty)];
+        if (_Execute_once(reinterpret_cast<once_flag&>(_Flag), _Immortalize_impl<_Ty>, &_Storage) == 0) {
```
Comment (StephanTLavavej) [inline stl/src/excptptr.cpp]: `_Flag` is already a `once_flag`.

File: tests/std/tests/VSO_0157762_feature_test_macros/env.lst [hunk 14]
```diff
@@ -1,7 +1,41 @@
 # Copyright (c) Microsoft Corporation.
 # SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
 
-RUNALL_INCLUDE ..\usual_matrix.lst
+#
+# This matrix is usual_matrix.lst but without /w14640 /Zc:threadSafeInit- on every possible line
```
Comment (StephanTLavavej) [inline tests/std/tests/VSO_0157762_feature_test_macros/env.lst]: It adds lines at the end, so this comment is inaccurate.

File: stl/src/excptptr.cpp (hunk 13 above)
Comment (BillyONeal) [reply stl/src/excptptr.cpp]: hahahaha

File: stl/inc/future (hunk 4 above)
Comment (BillyONeal) [reply stl/inc/future]: This is the kind of SHOUTY comment I don't like (it doesn't help in finding a pretty name) so will delete. (Same with the other ones)

File: stl/inc/system_error (hunk 1 above)
Comment (CaseyCarter) [reply stl/inc/system_error]: A core header cannot include `yvals.h`, but a non-core header may (and in fact certainly does indirectly) include `yvals_core.h`. When I added the `_STL_COMPILER_PREPROCESSOR` guards, I spammed
```c++
#include <yvals_core.h>
#if _STL_COMPILER_PREPROCESSOR
// ...
#endif // _STL_COMPILER_PREPROCESSOR
```
into all of the STL headers, and then cut-and-pasted any other `#include<yvals` line atop the newly-added `#include <yvals_core.h>`. We weren't concerned at the time with creating a "core/non-core headers always directly include yvals/yvals_core" invariant.

File: stl/inc/system_error [hunk 15]
```diff
@@ -520,16 +519,85 @@ public:
...
+// (there are probably others)
+//
+// Inspecting the resulting assembly of any callers of _Immortalize_memcpy_image is recommended.
+//
+
+#if defined(_M_CEE_PURE)
+// /clr:pure doesn't ever do constant initialization, so rely on the CLR and magic statics
+
+#pragma warning(push)
+#pragma warning(disable : 4640) // construction of local static object is not thread-safe (/Wall)
```
Comment (CaseyCarter) [inline stl/inc/system_error]: The comment on 554 says we're relying on magic statics, and the comment on 560 suggests that `_Static` is in fact a magic static. If those are correct, why would this warning be emitted necessitating that we suppress it?

File: stl/inc/system_error (hunk 1 above)
Comment (BillyONeal) [reply stl/inc/system_error]: I see, I'm going to leave this at yvals then.

File: stl/inc/system_error (hunk 15 above)
Comment (BillyONeal) [reply stl/inc/system_error]: Hmmmm.... good point. Let me try removing that and seeing what happens...

Comment (BillyONeal) [reply stl/inc/system_error]: Looks like that was unnecessary, removed.

File: stl/inc/system_error (hunk 8 above)
Comment (StephanTLavavej) [reply stl/inc/system_error]: Still active.

Comment (StephanTLavavej) [reply stl/inc/system_error]: Still active (although I am ok with wontfix here).

File: stl/inc/system_error (hunk 9 above)
Comment (StephanTLavavej) [reply stl/inc/system_error]: Still active (this doesn't meet the "super small preprocessor exemption").

**Comment by BillyONeal** [comment]
image

Exactly 3 things I did not address. Exactly 3 things GitHub hid. Thanks GitHub :(

**Pull request merged**